packages/astro/src/env/schema.ts

```diff
@@ -66,61 +108,32 @@ export const EnvSchema = z
 		}
 	});
 
+// https://www.totaltypescript.com/concepts/the-prettify-helper
+type Prettify<T> = {
+	[K in keyof T]: T[K];
+	// eslint-disable-next-line @typescript-eslint/ban-types
+} & {};
+
 export type EnvSchema = z.infer<typeof EnvSchema>;
 
-const StringField = z.intersection(
-	z.union([
-		PublicClientEnvFieldMetadata,
-		PublicServerEnvFieldMetadata,
-		SecretServerEnvFieldMetadata,
-	]),
-	StringSchema
-);
-export type StringField = z.infer<typeof StringField>;
-export const StringFieldInput = z.intersection(
-	z.union([
-		PublicClientEnvFieldMetadata,
-		PublicServerEnvFieldMetadata,
-		SecretServerEnvFieldMetadata,
-	]),
-	StringSchema.omit({ type: true })
-);
-export type StringFieldInput = z.infer<typeof StringFieldInput>;
+type _Field<T extends z.ZodType> = Prettify<z.infer<typeof EnvFieldMetadata & T>>;
+type _FieldInput<T extends z.ZodType, TKey extends string = 'type'> = Prettify<
+	z.infer<typeof EnvFieldMetadata> & Omit<z.infer<T>, TKey>
+>;
+
+export type StringField = _Field<typeof StringSchema>;
+export type StringFieldInput = _FieldInput<typeof StringSchema>;
+
+export type NumberField = _Field<typeof NumberSchema>;
+export type NumberFieldInput = _FieldInput<typeof NumberSchema>;
 
-const NumberField = z.intersection(
-	z.union([
-		PublicClientEnvFieldMetadata,
-		PublicServerEnvFieldMetadata,
-		SecretServerEnvFieldMetadata,
-	]),
-	NumberSchema
-);
-export type NumberField = z.infer<typeof NumberField>;
-export const NumberFieldInput = z.intersection(
-	z.union([
-		PublicClientEnvFieldMetadata,
-		PublicServerEnvFieldMetadata,
-		SecretServerEnvFieldMetadata,
-	]),
-	NumberSchema.omit({ type: true })
-);
-export type NumberFieldInput = z.infer<typeof NumberFieldInput>;
+export type BooleanField = _Field<typeof BooleanSchema>;
+export type BooleanFieldInput = _FieldInput<typeof BooleanSchema>;
 
-const BooleanField = z.intersection(
-	z.union([
-		PublicClientEnvFieldMetadata,
-		PublicServerEnvFieldMetadata,
-		SecretServerEnvFieldMetadata,
-	]),
-	BooleanSchema
-);
-export type BooleanField = z.infer<typeof BooleanField>;
-export const BooleanFieldInput = z.intersection(
-	z.union([
-		PublicClientEnvFieldMetadata,
-		PublicServerEnvFieldMetadata,
-		SecretServerEnvFieldMetadata,
-	]),
-	BooleanSchema.omit({ type: true })
-);
-export type BooleanFieldInput = z.infer<typeof BooleanFieldInput>;
+export type EnumField = _Field<typeof EnumSchema>;
+export type EnumFieldInput<T extends string> = Prettify<
+	_FieldInput<typeof EnumSchema, 'type' | 'values' | 'default'> & {
+		values: Array<T>;
+		default?: NoInfer<T> | undefined;
```

Comment:
This has been released in TS 5.4, could that be an issue?

Reply:
No, it should not